File: String/CPP/ExternalString.cpp
#include "ExternalString.hpp"

#include <stdio.h>
#include <stdlib.h>
#include <cstring>

namespace NumbatLogic
{
	bool ExternalString::Equal(const char* sxA, const char* sxB)
	{
		return strcmp(sxA, sxB) == 0;
	}

	// a bad checksum
	int ExternalString::GetChecksum(const char* sxString)
	{
		int nResult = 0xABC123;
		const char* p = sxString;
		while (*p)
			nResult = (nResult ^ *(p++)) << 1;
		return nResult;
	}

	long ExternalString::hextol(const char* sxString)
	{
		return ::strtol(sxString, NULL, 16);
	}

	int ExternalString::atoi(const char* sxString)
	{
		return ::atoi(sxString);
	}

	double ExternalString::atof(const char* sxString)
	{
		return ::atof(sxString);
	}

	long ExternalString::atol(const char* sxString)
	{
		return ::atol(sxString);
	}
}

File: String/CS/ExternalString.cs
namespace NumbatLogic
{
	class ExternalString
	{
		public static bool Equal(string sxA, string sxB)
		{
			return string.Equals(sxA, sxB);
		}

		public static int GetChecksum(string sxString)
		{
			int nResult = 0xABC123;
			for (int i = 0; i < sxString.Length; i++)
			{
				char c = sxString[i];
				nResult = (nResult ^ c) << 1;
			}
			return nResult;
		}

		public static long hextol(string sxString)
		{
			return long.Parse(sxString, System.Globalization.NumberStyles.HexNumber);
		}

		public static int atoi(string sxString)
		{
			return int.Parse(sxString);
		}

		public static long atol(string sxString)
		{
			return long.Parse(sxString);
		}

		public static double atof(string sxString)
		{
			return double.Parse(sxString);
		}
	}
}